### ISSUE TITLE
Consider relicensing to BSD/Apache/etc

### ISSUE THREAD

**Issue by bitprophet** [paramiko/paramiko#540]
tl;dr I've gotten a handful of requests to relicense Paramiko to a more permissive license.

At least a few each year since taking over in 2012, often from individuals representing nontrivial groups of commercial users from well known companies).

I've been tracking a bunch of notes about this in an Evernote but it really should be in a ticket for better visibility, discussion (ideally, not bikeshedding...) etc. Below is a dump of that Evernote and I'll clean it up in a bit.

----

* General resource: https://aprescott.com/posts/open-source-copyright
* http://www.jbkempf.com/blog/post/2012/How-to-properly-relicense-a-large-open-source-project - mentions how VLC is **uncommon** in how it explicitly retains license/copyright/etc for all contributors - implying most other projects do *not* do so.
* http://www.rosenlaw.com/pdf-files/Rosen_Ch10.pdf - says that the law is unambiguous and that we need permission.
* One of the folks asking for a relicense suggested a dual license but I assume that only works on a time split (code before X is license A, code after X is license B or A+B) and it'd be a very long time until enough code got rewritten/added before the LGPL parts "fell off".
    * Ditto re a general "we rewrote enough of this that we can safely claim the old stuff is gone" approach.
    * Unless the Py3 merge suffices, which I doubt - it modifies a lot of code but is not really rewriting it
* http://blog.gerv.net/2013/02/relicensing-when-do-you-have-to-ask/ - more evidence for "you must ask unless the licenses are compatible", so unless LGPL is compatible with BSD (highly unlikely, but I should check) that's evidence against.
* http://blog.yt-project.org/post/Relicensing.html - good discussion of a GPL -> BSD transformation. However it involved...getting permission from every contributor. They had 47. I assume Paramiko's got way more but should doublecheck.
    * Mentions that IPython did LGPL -> BSD, finding that email thread would be grand
    * Seems to be here? http://mail.scipy.org/pipermail/ipython-user/2004-October/002324.html
        * Interestingly, is taking a "I contacted you, silence is implied to be agreement" approach, inverse of the yt-project approach (kinda)
    * Also turns out Paramiko, going by Git(hub) only has 40 contributors, less than I was expecting given its age.
        * Robey didn't seem to keep an AUTHORS file or other attribution besides a single-digit number of names (no emails) given in ChangeLog files. So git history is all I really have.
        * Unclear how legally culpable we are for untrackable attributions like that, where we know it was not Robey but we can't get in touch.
        * Robey himself has granted me explicit permission to relicense, though he also raised the "but you probably need permission from everybody else too" point.
* http://lwn.net/Articles/247872/ - general discussion about the kinda-sorta relicense of ath5k from BSD to GPL
* http://www.jbkempf.com/blog/post/2012/How-to-properly-relicense-a-large-open-source-project - guide to relicensing via contacting contributors. **Have not read much yet.**
* An email discussion with one Christopher Egner yields these examples of OSS projects handling dual licenses or license switching:
    * http://blog.jquery.com/2012/09/10/jquery-licensing-changes/ - used to be dual GPL+MIT, dropped GPL
    * http://www.sqlite.org/copyright.html - public domain, but w/ option for a commercial license. (Doesn't really apply? Nobody has asked for a commercial Paramiko license that wouldn't be OK with simply a more-permissible OSS one.)

**Comment by bitprophet** [comment]
One pertinent excerpt from some of the email discussions I've had (this is from myself, to the other party):

> My and many others' interpretations of the GPL and LGPL is that
> Python's import statement does not count as "linking" in the sense the
> *GPL licenses use the term. Thus it's most likely safe to 'import
> some_gpl_library' within a more-permissively-licensed codebase,
> without 'tainting' that codebase & making it subject to the GPL.

That discussion was about a *more* permissive license (e.g. importing Paramiko into a BSD codebase), but AFAIK, the argument cuts both ways.

**Comment by lndbrg** [comment]
I actually disagree (and so do others: https://wiki.creativecommons.org/wiki/GPL_compatibility_use_cases#2._Using_Python_scripts_in_blender). Using a library that is GPL taints your whole project (unless it's LGPL which explicitly allows this).

I also got my contribution backed out from chef/ohai for adding a requirement to a GPLed rubygem. That might be an issue with them actually distributing all the gems as one big install though.

**Comment by bitprophet** [comment]
@lndbrg Well, so you're arguing that the above excerpt is incorrect about how it lumps GPL and LGPL together? Sure, I can see that. However, for our purposes, Paramiko *is* LGPL so the advice still stands, yes?

**Comment by lndbrg** [comment]
Yes, I argue that the lumping is incorrect. Given that we are using the LGPL we are not currently tainting any other product that uses paramiko (since linking/library usage is okay with proprietary software).

**Comment by bitprophet** [comment]
Yes, correct. I still feel it'd be worthwhile relicensing to BSD sometime; even though our reading of LGPL implies commercial use isn't a problem, it's clearly contentious/confusing enough that the issue comes up periodically :)

**Comment by theonewolf** [comment]
I'd like to chime in that from a commercial standpoint, this specific clause is a bit worrying:

>   6. As an exception to the Sections above, you may also combine or
> link a "work that uses the Library" with the Library to produce a
> work containing portions of the Library, and distribute that work
> under terms of your choice, provided that the terms permit
> modification of the work for the customer's own use and reverse
> engineering for debugging such modifications.

Combinations are defined in the Pythonic import sense I believe:

>   When a program is linked with a library, whether statically or using
> a shared library, the combination of the two is legally speaking a
> combined work, a derivative of the original library.  The ordinary
> General Public License therefore permits such linking only if the
> entire combination fits its criteria of freedom.  The Lesser General
> Public License permits more lax criteria for linking other code with
> the library.

Commercially speaking, you typically don't want to give "reverse engineering" rights to your products.

But the LGPL, according to my reading, requires this.

For example, lets say I wanted to package up a virtualenv inside an Python-executable Zip file and distribute that onto my customer's machines.

This would mean I'm distributing Paramiko which dynamically links into my commercial product.

This means, again according to my LGPL reading, that I must allow my customers the right to "reverse engineer" the combined work---my commercial product.

So, I may not need to open source my commercial bits, but I must not forbid reverse engineering of them.

I give a huge thumbs up towards switching to a more permissive license like the BSD license which you mentioned.

If you want to see what it could look like, check out oh-my-fish: https://github.com/oh-my-fish/oh-my-fish-legacy/pull/159

**Comment by ran-z** [comment]
Any update on this?

**Comment by bitprophet** [comment]
No comments, no update :) still in "may have to get buy-in from hundreds of previous contributors" limbo, and there's frankly many more pressing things on the priority list :(

**Comment by aihuasxy** [comment]
Paramiko can be used in closed source software and packaged by pyinstaller?
not modified, just package.

**Comment by bitprophet** [comment]
@aihuasxy As the current primary copyright holder I have, previously, given verbal permission for corporations to distribute unmodified copies of Paramiko, since the question of whether simple importing in Python triggers "derivative work" clauses of the LGPL is not clear cut.

**Comment by isamaru** [comment]
One additional concern is that "demos" examples are part of the project and included in the LGPL license. So if I use paramiko by modifying those, it seems to fall under derivative work and I "taint" my project.

**Comment by bitprophet** [comment]
@isamaru see #1245 😆 your comment is just another reason for me to wonder about whether they're truly useful...

@piskvorky it's all in this ticket! there is no behind the scenes work on this presently. basically, https://github.com/paramiko/paramiko/issues/540#issuecomment-305624710 🙃

**Comment by tony** [comment]
The license here surprised me.

To prevent the list from growing moving forward:

- Require a CLA for pull requests (suhh as https://github.com/cla-assistant/cla-assistant) and/or 
- have future commits license under BSD/MIT/ISC/etc.

> and there's frankly many more pressing things on the priority list :(

It's true.  I still think it's worth it to have a permissive license if that's the intent of the project.

Is this effort still worth it to you @bitprophet ?

**Comment by nodreb-borrus** [comment]
Hello, I am looking for some advice on how to use and modify this code:  https://github.com/paramiko/paramiko/blob/main/demos/forward.py

My project is licensed Apache 2.0, and the code above is LGPL.

This is an example script, not an import so I will be copying then modifying this file and including it in my project.

The license notice on the file lists the author's name and states it is a part of paramiko.  I will be modifying the file enough that this will not be true (my modified file is not included in paramiko).

I am fine leaving the notice as-is and including a copy of LGPL in my project if that is appropriate, but it doesn't seem right to blame the original author for any changes I make to the file.

Can anyone advise on what is appropriate to do here?

**Comment by bskinn** [comment]
@nodreb-borrus While licensing related, your question really is off-topic for this issue. If nothing else, this should be its own separate ticket.

That said, I don't know if you're likely to get many useful replies here. You're probably better asking on one of the StackExchange sites, such as https://opensource.stackexchange.com.

**Comment by WillFM** [comment]
Instead of waiting for explicit approval from all previous contributors. Could you instead do an explicit decline/protest?. with a deadline, And explicitly call out, that by not responding within the deadline, they are implicitly agreeing to the license change? That way, at least the Burdon, is not chasing down every single response. 

Large cooperation's do this all the time with changes to there licensing agreements.

You'd still be on the hook for sending some type correspondence to all contributors though.

**Comment by remram44** [comment]
There is no "implicit agreeing" to losing your copyright, wtf?
